source/slang/slang-check-decl.cpp

```diff
@@ -7261,12 +7307,172 @@ namespace Slang
+            auto tmpExpr = stmt->body;
+            auto seqStmt = m_astBuilder->create<SeqStmt>();
+            seqStmt->stmts.add(tmpExpr);
+            stmt->body = seqStmt;
+            return seqStmt;
+        }
+        return as<SeqStmt>(stmt->body);
+    }
+
+    static List<ConstructorDecl*> _getCtorList(StructDecl* structDecl, ConstructorDecl** defaultCtorOut)
+    {
+        List<ConstructorDecl*> ctorList;
+        for (auto m : structDecl->members)
+        {
+            auto ctor = as<ConstructorDecl>(m);
+            if (!ctor || !ctor->body)
+                continue;
+            ctorList.add(ctor);
+            if (ctor->members.getCount() != 0)
+                continue;
+            *defaultCtorOut = ctor;
+        }
+        return ctorList;
+    }
+
     void SemanticsDeclBodyVisitor::visitAggTypeDecl(AggTypeDecl* aggTypeDecl)
     {
         if (aggTypeDecl->hasTag(TypeTag::Incomplete) && aggTypeDecl->hasModifier<HLSLExportModifier>())
         {
             getSink()->diagnose(aggTypeDecl->loc, Diagnostics::cannotExportIncompleteType, aggTypeDecl);
         }
+    
+        auto structDecl = as<StructDecl>(aggTypeDecl);
+        if (!structDecl)
+            return;
+
+        struct DeclAndCtorInfo
+        {
+            StructDecl* parent = nullptr;
+            ConstructorDecl* defaultCtor = nullptr;
+            List<ConstructorDecl*> ctorList;
+            DeclAndCtorInfo()
+            {
+            }
+            DeclAndCtorInfo(StructDecl* parent, const bool getOnlyDefault)
+            {
+                if (getOnlyDefault)
+                    defaultCtor = _getDefaultCtor(parent);
+                else
+                    ctorList = _getCtorList(parent, &defaultCtor);
+            }
+        };
+        List<DeclAndCtorInfo> inheritanceDefaultCtorList{};
+        for (auto member : structDecl->members)
+        {
+            auto inheritanceMember = as<InheritanceDecl>(member);
+            if (!inheritanceMember)
+                continue;
+            auto declRefType = as<DeclRefType>(inheritanceMember->base.type);
+            if (!declRefType)
+                continue;
+            auto structOfInheritance = as<StructDecl>(declRefType->getDeclRef().getDecl());
+            if (!structOfInheritance)
+                continue;
+            inheritanceDefaultCtorList.add(DeclAndCtorInfo(structOfInheritance, true));
+        }
+        DeclAndCtorInfo structDeclInfo = DeclAndCtorInfo(structDecl, false);
+
+        std::size_t insertOffset = 0;
+        Dictionary<Decl*, Expr*> cachedDeclToCheckedVar;
+        for (auto& declInfo : inheritanceDefaultCtorList)
+        {
+            if (!declInfo.defaultCtor)
+                continue;
+            for (auto ctor : structDeclInfo.ctorList)
+            {
+                auto seqStmt = _ensureCtorBodyIsSeqStmt(m_astBuilder, ctor);
+                auto ctorName = declInfo.defaultCtor->getName();
+
+                auto lookupResult = lookUpMember(
```

Comment:
Isn't this just going to return ctor? In that case why not just create a VarRxpr whose DeclRef is ctor?

Seems like the two loops here should swap? Loop on every ctor, and inside it loop on every base?

Reply:
If we swap the loop, we will do more `loop`s despite computing the same work.
The current layout allows `if (!declInfo.defaultCtor) continue;` to skip an inner loop.

Reply:
I understand why the following was suggested, it was in reference to: 
> Avoid insert this many times. Create a single SeqStmt to hold all new stmts and simply insert that seqStmt at top.

Reply:
I will change the order, it will be faster to create 1 seqStmt then insert twice probably